articles/20240820_gpu_utilisation.md

```diff
@@ -0,0 +1,294 @@
+
+1. **Set Up the Package Repository**:
+
+   ```bash
+   distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
+   curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
+   curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
+   ```
+
+2. **Install the NVIDIA Container Toolkit**:
+
+   ```bash
+   sudo apt-get update
+   sudo apt-get install -y nvidia-container-toolkit
+   sudo systemctl restart docker
+   ```
+
+3. **Verify GPU Access in Docker**:
+
+   Run the following command to ensure Docker can access your GPU:
+
+   ```bash
+   docker run --gpus all nvidia/cuda:11.0-base nvidia-smi
+   ```
+
+   This should display the details of your GPU, confirming that Docker is configured correctly.
+
+### Configuring Docker for GPU Access
+
+Once Docker and the NVIDIA toolkit are set up, the next step is to configure a Dockerfile that leverages GPU resources. The Dockerfile is essentially a blueprint for creating Docker images, defining the environment in which your application will run.
+
+#### Example Dockerfile
+
+Here’s an example Dockerfile designed to set up a Python environment with GPU support. This configuration is suitable for running machine learning models that require heavy computational resources:
+
+```Dockerfile
+# Use an official NVIDIA CUDA image as a base
+FROM nvidia/cuda:11.3.1-cudnn8-devel-ubuntu20.04
+
+# Install Python and pip
+RUN apt-get update && apt-get install -y python3 python3-pip
+
+# Install necessary Python packages
+RUN pip3 install torch torchvision torchaudio
+
+# Set the working directory
+WORKDIR /app
+
+# Copy your application code into the container
+COPY . /app
+
+# Command to run your application
+CMD ["python3", "your_script.py"]
+```
+
+##### Breakdown of the Dockerfile
+
+- **Base Image**: We start with an NVIDIA CUDA base image that includes development libraries and tools (`nvidia/cuda:11.3.1-cudnn8-devel-ubuntu20.04`).
+- **Python Installation**: The `apt-get` commands install Python and pip.
+- **Python Packages**: The `pip3` commands install the necessary Python libraries, such as PyTorch, which supports GPU acceleration.
+- **Working Directory**: The `WORKDIR` command sets the working directory within the container.
+- **Application Code**: The `COPY` command transfers your application code into the container.
+- **Execution Command**: The `CMD` command specifies the script to run when the container starts.
+
+### Demo Project: Running nanoGPT in Daytona
+
+To demonstrate how to utilize GPUs in containerized environments, we'll use [nanoGPT](https://github.com/karpathy/nanoGPT), a lightweight implementation of the GPT model by Andrej Karpathy. This project is ideal for developers experimenting with LLM inference in containers.
+
+#### Cloning the nanoGPT Repository
+
+Start by cloning the nanoGPT repository from GitHub:
+
```

Comment:
When using Daytona, daytona is cloning the repo adn creating workspace out of it inside the docker container